scripts/make_option_enum.py

```diff
@@ -0,0 +1,214 @@
+                    '   {{\n'
+                    '      out = {};\n'
+                    '      return(true);\n'
+                    '   }}\n'.format(a, enum_value(enum, v)))
+
+        out.write(
+            '   else\n'
+            '   {\n'
+            '      return(false);\n'
+            '   }\n'
+            '}\n\n')
+
+    for enum in enums.values():
+        out.write(header)
+        out.write('const char *to_string({} val)\n'.format(enum.name))
+        out.write('{\n'
+                  '   switch (val)\n'
+                  '   {\n')
+
+        for v in enum.values:
+            vs = v if enum.convert_internal else v.lower()
+            out.write(
+                '   case {}:\n'
+                '      return "{}";\n\n'.format(
+                    enum_value(enum, v), vs))
+
+        out.write(
+            '   default:\n'
+            '      fprintf(stderr, "%s: Unknown {} \'%d\'\\n",\n'
+            '              __func__, static_cast<int>(val));\n'
+            '      log_flush(true);\n'
+            '      exit(EX_SOFTWARE);\n'
+            '   }}\n'
+            '}}\n\n'.format(enum.name))
+
+
+# -----------------------------------------------------------------------------
+def main():
+    parser = argparse.ArgumentParser(description='Generate options.cpp')
+    parser.add_argument('output', type=str,
+                        help='location of options.cpp to write')
+    parser.add_argument('header', type=str,
+                        help='location of options.h to read')
+    parser.add_argument('template', type=str,
+                        help='location of option_enum.cpp.in '
+                             'to use as template')
+    args = parser.parse_args()
+
+    with open(args.header, 'rt') as f:
+        for line in iter(f.readline, b''):
+            line = line.strip()
+
+            me = re_enum_decl.match(line)
+            if me is not None:
+                e = Enumeration(me.group(1), me.group(3), f)
+                enums[e.name] = e
+                continue
+
+            mv = re_values.match(line)
+            if mv is not None:
+                enum_name = mv.group(1)
+                enum = enums['{}_e'.format(enum_name)]
+                values[enum_name] = enum.values
+
+            ma = re_aliases.match(line)
+            if ma is not None:
+                alias_args = [x.strip() for x in ma.group(1).split(',')]
+                enum = enums[alias_args[0]]
+                enum.add_aliases(*alias_args[1:])
+
+    replacements = {
+        '##BANNER##': write_banner,
+        '##VALUE_STRINGS##': write_value_strings,
+        '##ALIASES##': write_aliases,
+        '##CONVERSIONS##': write_conversions,
+    }
+
+    with open(args.output, 'wt') as out:
+        with open(args.template, 'rt') as t:
+            for line in t:
+                directive = line.strip()
+                if directive in replacements:
+                    replacements[directive](out, args)
```

Comment:
I wish we could have used a real template engine but the standard one is too basic for what you are doing here.